torch/testing/_internal/autograd_function_db.py

```diff
@@ -0,0 +1,250 @@
+import torch
+from functools import partial
+from torch.testing import make_tensor
+from torch.testing._internal.opinfo.core import (
+    OpInfo,
+    SampleInput,
+)
+from torch.testing._internal.common_dtype import all_types_and
+import numpy as np
+
+# Note: [autograd.Function db]
+#
+# This is a collection of autograd.Function test cases written as OpInfos
+# so they can easily be consumed by OpInfo-based tests to check if a subsystem
+# supports autograd.Function.
+#
+# Axes:
+# - saves {output, input, intermediate, non-tensor}
+# - {inputs, outputs} x {single tensor, tensors, arbitrary objects}
+# - Uses {mark_dirty, mark_non_differentiable, once_differentiable}
```

Comment:
Very comprehensive testing!

For even greater completeness, we could add uses of `ctx.needs_input_grad` and `set_materialize_grads`.

Reply:
Good catch, will add

Reply:
I added some manual tests for these. They don't really affect correctness, so adding to autograd_function_db wasn't sufficient.